src/ppx_deriving_yojson.ml

```diff
@@ -38,69 +38,74 @@ let parse_options options =
     | _ -> raise_errorf ~loc:expr.pexp_loc "%s does not support option %s" deriver name);
   !strict
 
+let wrap_runtime decls =
+  [%expr let open Ppx_deriving_yojson_runtime in [%e decls]]
+
 let rec ser_expr_of_typ typ =
   let attr_int_encoding typ =
     match attr_int_encoding typ with `String -> "String" | `Int -> "Intlit"
   in
-  match typ with
-  | [%type: int]             -> [%expr fun x -> `Int x]
-  | [%type: float]           -> [%expr fun x -> `Float x]
-  | [%type: bool]            -> [%expr fun x -> `Bool x]
-  | [%type: string]          -> [%expr fun x -> `String x]
-  | [%type: bytes]           -> [%expr fun x -> `String (Bytes.to_string x)]
-  | [%type: char]            -> [%expr fun x -> `String (String.make 1 x)]
-  | [%type: [%t? typ] ref]   -> [%expr fun x -> [%e ser_expr_of_typ typ] !x]
-  | [%type: [%t? typ] list]  -> [%expr fun x -> `List (List.map [%e ser_expr_of_typ typ] x)]
-  | [%type: int32] | [%type: Int32.t] ->
-    [%expr fun x -> `Intlit (Int32.to_string x)]
-  | [%type: int64] | [%type: Int64.t] ->
-    [%expr fun x -> [%e Exp.variant (attr_int_encoding typ.ptyp_attributes)
-                                    (Some [%expr (Int64.to_string x)])]]
-  | [%type: nativeint] | [%type: Nativeint.t] ->
-    [%expr fun x -> [%e Exp.variant (attr_int_encoding typ.ptyp_attributes)
-                                    (Some [%expr (Nativeint.to_string x)])]]
-  | [%type: [%t? typ] array] ->
-    [%expr fun x -> `List (Array.to_list (Array.map [%e ser_expr_of_typ typ] x))]
-  | [%type: [%t? typ] option] ->
-    [%expr function None -> `Null | Some x -> [%e ser_expr_of_typ typ] x]
-  | [%type: Yojson.Safe.json] -> [%expr fun x -> x]
-  | { ptyp_desc = Ptyp_constr ({ txt = lid }, args) } ->
-    app (Exp.ident (mknoloc (Ppx_deriving.mangle_lid (`Suffix "to_yojson") lid)))
+  wrap_runtime begin
```

Comment:
You don't have to wrap every expression... in fact wrapping just the outermost function will suffice.